integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java

```diff
@@ -200,37 +242,54 @@ private void requestMasterContainer() throws Exception {
     if (!mMasterAddress.equals("localhost")) {
       relaxLocality = false;
     }
-    ContainerRequest masterContainerAsk =
-        new ContainerRequest(masterResource, nodes, null /* any racks */, PRIORITY, relaxLocality);
+    ContainerRequest masterContainerAsk = new ContainerRequest(masterResource, nodes,
+        null /* any racks */, MASTER_PRIORITY, relaxLocality);
     LOG.info("Making resource request for Tachyon master: cpu {} memory {} MB on node {}",
         masterResource.getVirtualCores(), masterResource.getMemory(), mMasterAddress);
     mRMClient.addContainerRequest(masterContainerAsk);
 
-    // Wait until Tachyon master container has been allocated
+    LOG.info("Waiting for master container to be allocated");
     mMasterContainerAllocatedLatch.await();
   }
 
   /**
-   * Requests containers for the workers and waits for them all to be allocated.
+   * Requests containers for the workers, attempting to get containers on separate nodes.
    */
   private void requestWorkerContainers() throws Exception {
+    LOG.info("Requesting worker containers");
     // Resource requirements for worker containers
     Resource workerResource = Records.newRecord(Resource.class);
     workerResource.setMemory(mWorkerMemInMB + mRamdiskMemInMB);
     workerResource.setVirtualCores(mWorkerCpu);
+    int currentNumWorkers;
+    synchronized (mWorkerHosts) {
+      currentNumWorkers = mWorkerHosts.size();
+    }
 
     // Make container requests for workers to ResourceManager
-    for (int i = 0; i < mNumWorkers; i ++) {
-      ContainerRequest containerAsk =
-          new ContainerRequest(workerResource, null /* any hosts */, null /* any racks */,
-              PRIORITY);
-      LOG.info("Making resource request for Tachyon worker {}: cpu {} memory {} MB on any nodes",
-          i, workerResource.getVirtualCores(), workerResource.getMemory());
+    for (int i = currentNumWorkers; i < mNumWorkers; i ++) {
+      ContainerRequest containerAsk = new ContainerRequest(workerResource, getUnusedWorkerHosts(),
```

Comment:
Good call, Yarn gets mad if we demand an empty array of host names. I've added a more general check which ensures there are enough unused nodes to host all requested workers